### PR TITLE
Add OpenChat, Alpaca, Vicuna chat templates

### PR DESCRIPTION
**Pull request by kaizau**
This PR adds chat templates for some of the more popular non-ChatML models (that I know of, at least!).

- [OpenChat](https://huggingface.co/openchat/openchat-3.5-0106) is also used by [Starling 7B](https://huggingface.co/Nexusflow/Starling-LM-7B-beta).
- [Vicuna 1.1](https://huggingface.co/eachadea/vicuna-13b-1.1) and its Orca variant are used by a bunch of Yi fine tunes ([Nous Capybara](https://huggingface.co/NousResearch/Nous-Capybara-34B), [Tess](https://huggingface.co/migtissera/Tess-34B-v1.5b), etc.)
- [Alpaca](https://huggingface.co/wxjiao/alpaca-7b) is used by [DeepSeek Coder](https://huggingface.co/deepseek-ai/deepseek-coder-33b-instruct/).

Named `openchat`, `vicuna`, and `alpaca` respectively.

I based OpenChat's on the official [Jinja template](https://huggingface.co/openchat/openchat-3.5-0106/blob/main/tokenizer_config.json#L51), and Vicuna's on the one from [text-generation-web-ui](https://github.com/oobabooga/text-generation-webui/tree/main/instruction-templates) (couldn't find it in any model's config_tokenizer.json, but it matches what I saw in model cards and HF discussions). Alpaca was done using DeepSeek's template since the original also predates Jinja chat templates.

Caveat: Because none of the Vicuna models I've tested seem to include a chat template string, there doesn't seem to be a good way to heuristically detect the Orca variant. I've worked around this by creating a `vicuna-orca` template that's also handled by `vicuna`. Open to alternatives here.

New to C++ and this project, so please don't hesitate to mention any details I may have missed!

### CODE REVIEW

File: llama.cpp
Comment (ngxson) [inline llama.cpp:15745]: alpaca template and deepseek template both look similar at the first glance, but the main different is that alpaca template only used for instruction-response (one turn) and not multiple turns like modern chat template.

deepseek extends the notion of instruction-response into multi-turn by placing `<|EOT|>` token between each turn, so the formatted chat should look like:

```
### Instruction:
who are you?
### Response:
I am assistant
<|EOT|>
### Instruction:
1+1 is
### Response:
equal to 2
<|EOT|>
```

So what missing here is that `<|EOT|>`

File: tests/test-chat-template.cpp
Comment (ngxson) [inline tests/test-chat-template.cpp:43]: We need chat template to be in `config_tokenizer.json` is because that's the only way convert scripts can fetch this information. If this template never appears in `config_tokenizer.json`, then llama.cpp can never detect them automatically, users will need to switch template manually.

Personally, I don't feel like having vicuna/alpaca as chat template is a good idea. The reason is because these models come from era when we only need the model to do single turn instruction-response, so technically they are not chat template.

**Review by ngxson — CHANGES_REQUESTED** [review]
Could you also test to see if the output really matches with the python version of these template? You can use the python code here: https://github.com/ggerganov/llama.cpp/wiki/Templates-supported-by-llama_chat_apply_template

Please also let me know what must be added to the wiki page.

File: llama.cpp
Comment (ngxson) [reply llama.cpp:15745]: The chat above is produced by python code + jinja template, it doesn't seem to have `"\n\n"` at the end of each message, so I think the `"\n\n"` should be replaced by `"\n"`

File: llama.cpp
Comment (Jeximo) [inline llama.cpp:15736]: Typo? `("### Instruction:\\n")`, looks like an extra \

**Comment by Jeximo** [comment]
Mistral Instruct may be good for templating, https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2

File: llama.cpp
Comment (kaizau) [reply llama.cpp:15745]: Thanks for the python script! Included the Jinja output of OpenChat and DeepSeek below. And as you mentioned, the other two fail due to not having templates in `config_tokenizers.json`.

Will add `<|EOT|>` for DeepSeek when I have moment tomorrow.

```
openchat/openchat-3.5-0106
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
<s>GPT4 Correct System: You are a helpful assistant<|end_of_turn|>GPT4 Correct User: Hello<|end_of_turn|>GPT4 Correct Assistant: Hi there<|end_of_turn|>GPT4 Correct User: Who are you<|end_of_turn|>GPT4 Correct Assistant:    I am an assistant   <|end_of_turn|>GPT4 Correct User: Another question<|end_of_turn|>
------------------------------
deepseek-ai/deepseek-coder-33b-instruct
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
<｜begin▁of▁sentence｜>You are a helpful assistant### Instruction:
Hello
### Response:
Hi there
<|EOT|>
### Instruction:
Who are you
### Response:
   I am an assistant   
<|EOT|>
### Instruction:
Another question

------------------------------
```

File: tests/test-chat-template.cpp
Comment (kaizau) [reply tests/test-chat-template.cpp:43]: Yeah, that makes sense. I included Vicuna and Alpaca not for those models, but newer ones that borrowed their chat formats (even if with tweaks).

Would you be open to having templates that aren't auto-detected, but only accessible from the server's `--chat-template`? Server is my main use case, so having to manually select isn't a big deal.

File: llama.cpp
Comment (kaizau) [reply llama.cpp:15736]: My understanding is that find works on the `chat_template` string, which at least in DeepSeek's case, escapes the newline: https://huggingface.co/deepseek-ai/deepseek-coder-33b-instruct/blob/main/tokenizer_config.json

But let me test it out tomorrow — planning to separate out DeepSeek out in any case.

Comment (Jeximo) [reply llama.cpp:15736]: Sorry, I think you're correct. Thanks.

**Comment by kaizau** [comment]
@ngxson Three updates and a question:

1. I tweaked the python script to match the history and `add_generation_prompt` settings from `tests/test_chat_template.cpp`. Also included a shortcut for copying the output as a test.
    <details>
    <summary>Updated script</summary>

    ```python
    from transformers import AutoTokenizer

    VARIANTS_TO_TEST = [
        #'teknium/OpenHermes-2.5-Mistral-7B',
        # 'mistralai/Mistral-7B-Instruct-v0.2',
        # 'TheBloke/FusionNet_34Bx2_MoE-AWQ',
        # 'bofenghuang/vigogne-2-70b-chat',
        # 'mlabonne/AlphaMonarch-7B',
        # 'google/gemma-7b-it',
        # 'OrionStarAI/Orion-14B-Chat',
        # 'openbmb/MiniCPM-2B-dpo-fp32',
        'openchat/openchat-3.5-0106',
        'deepseek-ai/deepseek-coder-33b-instruct',
    ]

    HISTORY = [
        { 'role': 'system', 'content': 'You are a helpful assistant' },
        { 'role': 'user', 'content': 'Hello' },
        { 'role': 'assistant', 'content': 'Hi there' },
        { 'role': 'user', 'content': 'Who are you' },
        { 'role': 'assistant', 'content': '   I am an assistant   ' },
        { 'role': 'user', 'content': 'Another question' },
    ]

    for variant in VARIANTS_TO_TEST:
        history = [m for m in HISTORY] # copy
        if 'Mistral' in variant or 'gemma' in variant:
            history.pop(0) # no system prompt for mistral and gemma
        if 'gemma' in variant:
            # GemmaTokenizer is quite buggy, let's hard code the template here
            GEMMA_TMLP = "{% if messages[0]['role'] == 'system' %}{{ raise_exception('System role not supported') }}{% endif %}{% for message in messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/user/assistant/...') }}{% endif %}{% if (message['role'] == 'assistant') %}{% set role = 'model' %}{% else %}{% set role = message['role'] %}{% endif %}{{ '<start_of_turn>' + role + '\n' + message['content'] | trim + '<end_of_turn>\n' }}{% endfor %}{% if add_generation_prompt %}{{'<start_of_turn>model\n'}}{% endif %}"
            print('[Gemma]')
            output = AutoTokenizer.from_pretrained(VARIANTS_TO_TEST[0]).apply_chat_template(history, tokenize=False, add_generation_prompt=True, chat_template=GEMMA_TMLP)
            print(output)
            print(output.replace("\n", "\\n"))
            print('-' * 30)
        else:
            print('[' + variant + ']')
            tokenizer = AutoTokenizer.from_pretrained(variant)
            output = tokenizer.apply_chat_template(history, tokenize=False, add_generation_prompt=True)
            print(output)
            print(output.replace("\n", "\\n"))
            print('-' * 30)
    ```
    </details>

2. Used this output as the ground truth tests for updating the templates. And yup, found differences — fixed them.

3. Replaced `alpaca` entirely with `deepseek`, re: your comment.

Question is about `vicuna` and `vicuna-orca`, and more generally, any models where automated detection isn't feasible. Would it make sense to support them if only through the `--chat-template` server flag? Or would you prefer I just cut them from this PR — maybe try to figure out an alternative later?

**Comment by Jeximo** [comment]
> Would it make sense to support them if only through the --chat-template server flag?

[When I search Orca](https://huggingface.co/microsoft/Orca-2-7b), it's `--chat-ml`, but I also saw the older-styled templates for Nous/Tess. I like **more options** for chat templates personally, but I understand not wanting to [complicate other development](https://github.com/ggerganov/llama.cpp/issues/6391#issue-2216074620), so leave it up to yourself and @ngxson.

File: llama.cpp
Comment (ngxson) [reply llama.cpp]: Looks OK to me, just a quick note is that `<｜begin▁of▁sentence｜>` is not needed, because BOS is always added on server

**Comment by ngxson** [comment]
Thanks for the efforts @kaizau . IMO chat/prompt template has always been a quite messy topic (rabbit hole as you said). You can see on the beginning of https://github.com/ggerganov/llama.cpp/issues/4216 there was a discuss about that.

After some more researches I think it's OK to keep vicuna/vicuna-orca. While they does not have official jinja template, I think we can maybe ask the model's author to add one (or the one to convert it to gguf to add one). One of the thing I fear was that some template does not have multi-turn capability from the beginning, like alpaca for example, but people try to retro-fit it. Turns out, that's not the case of vicuna, so it's safe to assume that all vicuna-based models support multi-turn.

**Comment by kaizau** [comment]
@ngxson Makes sense. Any other code / formatting changes you'd like to see here? I'll draft up a readme update shortly.

Relatedly, a quirk I've noticed in using the OpenChat and Vicuna templates is that the first character of every assistant message is now always " ".

This is because these 3 templates all use ": " as the role separator — yet all of the official / reference `add_generation_prompt` examples exclude the space after the colon.

I can't tell if this is an oversight or an intended. Adding the space after the colon in each `add_ass` block gets rid of the problem — which is what I would lean towards, and what I expect most users would prefer.

Did you encounter anything similar with previous chat templates?

**Comment by Jeximo** [comment]
> first character of every assistant message is now always " "

[readme under](https://github.com/p-e-w/llama.cpp/blob/master/examples/server/README.md#api-endpoints) `prompt` states *If the prompt is a string or an array with the first element given as a string, a bos token is inserted in the front like main does*

I'm not sure of the correct solution - I had a similar experience with CLI in `main`: `--in-prefix "GPT4 Correct User: " --in-suffix "<|end_of_turn|>GPT4 Correct Assistant:"`

I included the space for User, and **excluded** it for Assistant in order to strickly adhere to the template. I think it's intentional, but I may be wrong.

**Comment by kaizau** [comment]
@ngxson Was about to paste the readme update here, but realized I already had edit access to the page?

Either way, added the 4 templates: https://github.com/ggerganov/llama.cpp/wiki/Templates-supported-by-llama_chat_apply_template

I also added a "how to add a template" section that hopefully makes it incrementally easier for others to add templates. It includes the updated version of your script that makes it easier to copy-and-paste the python output as a test.

**Comment by SamuelTallet** [comment]
@kaizau 
The documentation part `<s>GPT4 Correct System: ` seems incorrect.

The system prompt shoud be appended without prefix.

Source: https://huggingface.co/openchat/openchat_3.5/discussions/5#65448109b4a3f3a2f486fd9d

**Comment by ngxson** [comment]
> Relatedly, a quirk I've noticed in using the OpenChat and Vicuna templates is that the first character of every assistant message is now always " ".

That's because tokenizers tend to encode both the word and the space into the same token. For example, using https://platform.openai.com/tokenizer :

image

Adding a trailing space in the assistant prompt `GPT4 Correct System: ` will make the model to perceive the sentence differently, because now the trailing space is encoded as a single character and not attached to the next word:

image

Sadly there's no other way to get rid of this problem. The root cause in fact is because this class of template does not have special tokens like `<|im_start|>`, they rely on common characters like `:` or space which is dependent on the next word.

**Comment by ngxson** [comment]
> I also added a "how to add a template" section that hopefully makes it incrementally easier for others. It includes the updated version of your script that outputs in a format identical to test-chat-template.cpp, to reduce room for human error.

Nice, thanks! That looks good to me. I don't know how the permission system in wiki page works, but I glad to know that you have write access to wiki.

File: llama.cpp
Comment (ngxson) [inline llama.cpp:15739]: The only thing left need to do is to remove this `<｜begin▁of▁sentence｜>`, this is because server already add BOS token to input prompt by default.

**Review by ngxson — CHANGES_REQUESTED** [review]

**Comment by ggerganov** [comment]
> I don't know how the permission system in wiki page works, but I glad to know that you have write access to wiki.

It might be a good idea to restrict wiki access to collaborators, agree?

image

**Comment by kaizau** [comment]
> The part `<s>GPT4 Correct System: ` in the Wiki seems incorrect.
> 
> OpenChat author said the system prompt should be appended without prefix.
> 
> Source: https://huggingface.co/openchat/openchat_3.5/discussions/5#65448109b4a3f3a2f486fd9d

@SamuelTallet Thanks! I saw that thread too and originally implemented the unprefixed version. 

But running the [actual Jinja template](https://huggingface.co/openchat/openchat_3.5/blob/main/tokenizer_config.json#L51) from the model's tokenizer_config.json produces `<s>GPT4 Correct System: `. So any implementation that actually uses the Jinja template _would_ include a prefix...The [readme](https://github.com/imoneoi/openchat) also references the tokenizer.chat_template as the correct one.

This is unfortunately the state of templates right now. 🥲

I've [left a comment](https://huggingface.co/openchat/openchat_3.5/discussions/45) asking for clarification, but will default to the unprefixed.

**Comment by ngxson** [comment]
@ggerganov Yes, it is important to restrict write access to wiki. Ideally IMO we can allow only a list of people (not all contributors), but I'm not sure if this option is possible on github. The reason is because changes to wiki does not requires review.

**Review by ngxson — APPROVED** [review]
LGTM. Thank you!

**Comment by ggerganov** [comment]
Updated the wiki access:

image

Note, these are only the collaborators that have write access (i.e. not all contributors). Still, if we want to make this even stricter, it should be moved as doc files and committed to the repo

**Comment by Folko-Ven** [comment]
@kaizau Hello, I apologize for disturbing you, but is there any hope for the addition of Mistral templates?

**Comment by ngxson** [comment]
@ggerganov Thanks. That's ok for now I think. We can consider moving wiki to doc files later. Personally, I still feel like the UI of wiki page is more simple to navigate.

@Folko-Ven Mistral uses llama2 template. Maybe we can add `mistral` as an alias for `llama2` to clarify that.

**Comment by Folko-Ven** [comment]
@ngxson It seemed to me that they are slightly different, aren’t they? Usually, I look at chat templates here - [[link]](https://github.com/chujiezheng/chat_templates)

**Comment by ngxson** [comment]
We do support 3 variants of llama2. Mistral uses the variant with spaces around message content. As long as the model have the correct jinja template, it will be auto-detected and correct template will be used.

**Comment by Folko-Ven** [comment]
@ngxson Got it, thanks for explaining!

**Comment by wtarreau** [comment]
Regarding the limitations to access the wiki from a list of people, the only solution we've found in haproxy was to create a dedicated project for the wiki and send invites to those who want to contribute. The main project's wiki is simply redirected to the wiki project and that solved the issues. but it's indeed annoying.

**Pull request merged**
